File: LeetCode/236. 二叉树的最近公共祖先.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function (root, p, q) {
  let s = [];
  let visited = new Set();
  let candidates = new Set();
  s.push(root);
  let idx = 0;
  while (s.length) {
    let cur = s.pop();
    if (!visited.has(cur)) {
      visited.add(cur);
      if (cur.right) s.push(cur.right);
      s.push(cur);
      if (cur.left) s.push(cur.left);
    } else {
      if (cur === p || cur === q) {
        idx++;
      }
      if (idx === 1) {
        candidates.add(cur);
      }
      if (idx === 2) {
        candidates.add(cur);
        break;
      }
    }
  }
  s = [];
  visited = new Set();
  s.push(root);
  while (s.length) {
    let cur = s.pop();
    if (!visited.has(cur)) {
      visited.add(cur);
      if (cur.right) s.push(cur.right);
      if (cur.left) s.push(cur.left);
      s.push(cur);
    } else {
      if (candidates.has(cur)) return cur;
    }
  }
  return null;
};

File: LeetCode/236. 二叉树的最近公共祖先.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn lowest_common_ancestor(root: Option<Rc<RefCell<TreeNode>>>, p: Option<Rc<RefCell<TreeNode>>>, q: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        if root == None {
            return None;
        }
        if root == p {
            return root.clone();
        }
        if root == q {
            return root.clone();
        }
        let root = root.unwrap();
        let left = Self::lowest_common_ancestor(root.borrow().left.clone(), p.clone(), q.clone());
        let right = Self::lowest_common_ancestor(root.borrow().right.clone(), p.clone(), q.clone());
        match (left.clone(), right.clone()) {
            (Some(_), Some(_)) => Some(root.clone()),
            _ => if left.is_some() { left } else { right }
        }
    }
}
